alternator/executor.cc

```diff
@@ -96,6 +96,44 @@ std::string make_jsonable::to_json() const {
     return rjson::print(_value);
 }
 
+
+static bool is_leaf(const rjson::value& value) {
+    return !value.IsObject() && !value.IsArray();
+}
+
+static future<> destroy_gently_nonleaf(rjson::value&& value_in);
+
+future<> destroy_gently(rjson::value&& value) {
+    if (is_leaf(value)) {
+        return make_ready_future<>();
+    } else {
+        return destroy_gently_nonleaf(std::move(value));
+    }
+}
+
+static
+future<>
+destroy_gently_nonleaf(rjson::value&& value_in) {
+    // value_in is an rvalue ref to force caller to move
+    auto value = std::move(value_in);
+
+    if (value.IsObject()) {
+        for (auto it = value.MemberBegin(); it != value.MemberEnd();) {
+            // destroy_gently() is not a coroutine to avoid allocation if it->value is a leaf
+            co_await destroy_gently(std::move(it->value));
+            it = value.EraseMember(it);
+        }
+    } else if (value.IsArray()) {
+        for (auto i = value.Size(); i > 0; --i) {
+            auto index = i - 1;
+            if (is_leaf(value[index])) {
+                co_await destroy_gently(std::move(value[index]));
+                value.Erase(value.Begin() + index);
+            }
+        }
+    }
```

Comment:
I didn't understand why it wasn't enough to have one recursive function destroy_gently, and you needed that _nonleaf variant. This function could have easily handled both variants. Is it because of the "not a coroutine" statement in the comment above? Couldn't you have made the entire function **not** a coroutine, or you avoided that because it's harder to code an old-style continuation-based loop?

Reply:
There's a comment above:

// destroy_gently() is not a coroutine to avoid allocation if it->value is a leaf

Reply:
I could have made the whole thing not a coroutine, but I am not a masochist.